RUFAS/biophysical/animal/herd_manager.py

```diff
@@ -1,2 +1,2079 @@
+            AnimalType.HEIFER_II: self.heiferIIs,
+            AnimalType.HEIFER_III: self.heiferIIIs,
+            AnimalType.LAC_COW: [cow for cow in self.cows if cow.is_milking],
+            AnimalType.DRY_COW: [cow for cow in self.cows if not cow.is_milking],
+        }
+
+    @property
+    def animals_by_combination(self) -> dict[AnimalCombination, list[Animal]]:
+        """
+        Group animals by combination.
+
+        Returns
+        -------
+        dict[AnimalCombination, list[Animal]]
+            A dictionary where the keys are instances of `AnimalCombination`, and
+            the values are lists of `Animal` instances belonging to the corresponding
+            combination.
+
+        """
+        animals_by_combination = defaultdict(list)
+        for animal in [
+            *self.calves,
+            *self.heiferIs,
+            *self.heiferIIs,
+            *self.heiferIIIs,
+            *self.cows,
+        ]:
+            animal_combination = self.ANIMAL_GROUPING_SCENARIO.find_animal_combination(animal)
+            animals_by_combination[animal_combination].append(animal)
+        return animals_by_combination
+
+    @property
+    def pens_by_animal_combination(self) -> dict[AnimalCombination, list[Pen]]:
+        """
+        Group a list of pens by animal combination.
+
+        Returns
+        -------
+        Dict[AnimalCombination, list[Pen]]
+            Dictionary of pens grouped by animal combination.
+
+        """
+
+        pen_group_by_animal_combination = defaultdict(list)
+        for pen in self.all_pens:
+            pen_group_by_animal_combination[pen.animal_combination].append(pen)
+        return pen_group_by_animal_combination
+
+    @property
+    def phosphorus_concentration_by_animal_class(self) -> dict[AnimalType, float]:
+        """
+        Retrieves the phosphorus concentration for each animal type.
+
+        Returns
+        -------
+        dict[AnimalType, float]
+            A dictionary mapping each animal type to its corresponding phosphorus
+            concentration. If the total body weight of an animal class is zero, the
+            phosphorus concentration is set to 0.0 for that class.
+
+        Notes
+        -----
+        These variables are the P concentrations of each class of animal. They are calculated daily and are used when an
+        animal is added to the herd, whether by birth or replacement herd purchase. They  are calculated by dividing the
+        total P in the animals of the class by the total body weight of the animals, on a per-animal basis.
+
+        """
+        phosphorus_concentration_by_animal_class: dict[AnimalType, float] = {
+            animal_type: 0.0
+            for animal_type in [
+                AnimalType.CALF,
+                AnimalType.HEIFER_I,
+                AnimalType.HEIFER_II,
+                AnimalType.HEIFER_III,
+                AnimalType.LAC_COW,
+                AnimalType.DRY_COW,
+            ]
+        }
```

Comment:
```suggestion
        phosphorus_concentration_by_animal_class: dict[AnimalType, float] = {animal_type: 0.0 for animal_type in AnimalType
        }
```
I didn't know this before just now, but you can iterate directly through the members of an enum in Python